Cat Shop with Spring Security

markruler opened the pull request:
## 과제 목표

### 로그인이 필요없는 API

- [x] 로그인 - `POST /session`
- [x] 회원 생성하기 - `POST /users`
- [x] 고양이 장난감 목록 얻기 - `GET /products`
- [x] 고양이 장난감 상세 조회하기 - `GET /products/{id}`

### 로그인이 필요한 API

- [x] 고양이 장난감 등록하기 - `POST /products`
- [x] 고양이 장난감 수정하기 - `PATCH /products/{id}`
- [x] 고양이 장난감 삭제하기 - `DELETE /products/{id}`
- [x] 회원 수정하기 - `PATCH /users/{id}`
- [x] 회원 삭제하기 - `DELETE /users/{id}`

## 요구 사항

- [ ] 테스트 커버리지 100%를 달성해야 합니다.
- [x] 모든 API 테스트를 통과해야 합니다.
- [x] 모든 E2E 테스트를 통과해야 합니다.

markruler committed:
Add spring security for auths

- Authentication
- Authorization

markruler committed:
Add JavaDoc to auth-related classes

johngrib reviewed (commented):
이번주 부터는 커밋 메시지도 같이 연습해 봅시다. 커밋 메시지 타이틀과 커밋 메시지 본문을 한국어로 작성해서 올려주시면 커밋 메시지에 대해서도 리뷰해 드릴게요.

johngrib commented on app/build.gradle:
```diff
@@ -69,6 +69,9 @@ dependencies {
     runtimeOnly 'com.h2database:h2'
 
+    // Spring Security
+    implementation 'org.springframework.boot:spring-boot-starter-security'
```
스프링 시큐리티에 대해 조사하고, 동료들에게 간단하게 소개하는 글을 작성해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -5,30 +5,53 @@
+     * @param email    이메일
+     * @param password 비밀번호
+     * @return 토큰
+     * @throws LoginFailException 이메일을 찾을 수 없거나 비밀번호가 일치하지 않을 경우 던집니다.
```
이렇게 해도 충분할 것 같아요.
```suggestion
     * @throws LoginFailException 로그인에 실패하면 던집니다
```

johngrib commented on app/src/main/java/com/codesoom/assignment/config/PasswordEncoderJavaConfig.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * 패스워드 인코더 설정
+ */
+@Configuration
+public class PasswordEncoderJavaConfig {
+    @Bean
```
`@Configuration`과 `@Bean`의 JavaDoc을 읽고 두 애노테이션에 대해 간단하게 설명해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,46 @@
+        Filter authenticationErrorFilter = new AuthenticationErrorFilter();
+
+        //@formatter:off
+        http
+            .addFilter(authenicationFilter)
+                .addFilterBefore(authenticationErrorFilter, JwtAuthenticationFilter.class)
+            .csrf()
+                .disable()
+            .sessionManagement()
+                .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and()
+            .exceptionHandling()
+                .authenticationEntryPoint(new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED));
```
이 코드 한줄한줄이 어떤 의미를 갖는 설정인지 설명해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,72 @@
+        }
+
+        String method = request.getMethod();
+        if (method.equals("GET")) {
+            return true;
+        }
+
+        if (method.equals("OPTIONS")) {
```
`GET`, `OPTIONS` 같은 것들은 이미 만들어져 있는 enum이 어딘가 있지 않을까요?

markruler committed:
권한이 필요한 `/users` API에서 회원 ID가 같은지 검사한다

회원 정보를 수정하거나 삭제할 경우 요청한 사용자의 ID와 요청된 사용자 정보의 ID가 같은지 확인합니다. 특정 사용자가 다른 사용자 정보를 임의로 수정하거나 삭제하는 것을 방지합니다.

markruler committed:
JavaDoc을 수정한다

- AuthenticationService

johngrib commented on app/src/main/java/com/codesoom/assignment/errors/ForbiddenApiRequestException.java:
```diff
@@ -0,0 +1,7 @@
+package com.codesoom.assignment.errors;
+
+public class ForbiddenApiRequestException extends RuntimeException {
+    public ForbiddenApiRequestException() {
+        super("타인의 정보를 수정할 수 없습니다.");
+    }
+}
```
클래스 이름에 비해 에러 메시지의 커버 범위가 너무 좁습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -5,30 +5,53 @@
+     * @param email    이메일
+     * @param password 비밀번호
+     * @return 토큰
+     * @throws LoginFailException 로그인에 실패하면 던집니다
```
```suggestion
     * @throws LoginFailException 로그인에 실패했을 경우
```

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/ProductController.java:
```diff
@@ -40,16 +45,16 @@ public Product detail(@PathVariable Long id) {
     }
 
     @PatchMapping("{id}")
+    @PreAuthorize("isAuthenticated() and hasAuthority('USER')")
```
문자열로 주어진 값이 어떤 의미를 갖는지 설명해 보세요. `USER`는 어떤 뜻을 갖고 있나요? `USER` 대신 입력할 수 있는 값은 어떤어떤 것들이 있나요?

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserControllerErrorAdvice.java:
```diff
@@ -39,18 +37,12 @@ public ErrorResponse handleLoginFailException() {
     public ErrorResponse handleConstraintValidateError(
             ConstraintViolationException exception
-    ) {
+                                                      ) {
```
여기 인덴트가 너무 깊이 들어간 것 같아요.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/AuthenticationErrorFilter.java:
```diff
@@ -0,0 +1,28 @@
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class AuthenticationErrorFilter extends HttpFilter {
```
이 필터의 책임과 기능에 대해 JavaDoc으로 설명해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/JwtUtil.java:
```diff
@@ -18,13 +23,26 @@ public JwtUtil(@Value("${jwt.secret}") String secret) {
+     *
+     * @param token 토큰
+     * @return 토큰에 포함된 정보
+     * @throws InvalidTokenException 토큰이 빈 값이거나 시그니처가 올바르지 않다면 던집니다.
```
```suggestion
     * @throws InvalidTokenException 토큰이 빈 값이거나 시그니처가 올바르지 않을 경우
```

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 요즘 많이 바쁘셨나봐요. 힘내시길 바랍니다.
주간 회고를 작성해 공유해 주세요.

markruler committed:
ForbiddenApiRequestException의 메시지는 외부에서 입력하도록 한다

markruler committed:
JavaDoc의 `@throws` 설명을 수정한다

문장을 '~하다'가 아닌 '~한 경우'와 같이 끝맺습니다.

markruler committed:
불필요한 인덴트는 제거한다
